### PR TITLE
Closes #2102 Error upon save when linking to a document in a Card Link URL field

### LINKED ISSUE #2102: Error upon save when linking to a document in a Card Link URL field
## Problem/Motivation
When attempting to use the card link field to link to a document (PDF) on the site, you receive an error upon trying to save the page. Document links work fine in the card body, but not in the card link field.
Screen Shot 2022-12-16 at 11 33 33 AM

### PR DESCRIPTION
**Pull request by bberndt-uaz**
## Description

Allows documents (and any public file) to be linked in the Card Link URL field. If the link starts with the public files base path (which is what Linkit autocomplete uses for file links), the form validation will verify that the file exists.

Previous clarifying questions:

1. Is it OK to ignore validation for public files?
   - Users can already get around the validation by using absolute URLs.
   - If not, how could we check if a file exists at the URL?
   - (Update Nov 6, 2023) Validation added for file links to confirm that the files do exist.
2. Could/should we use the [Linkit field formatter](https://git.drupalcode.org/project/linkit/-/blob/6.1.x/src/Plugin/Field/FieldFormatter/LinkitFormatter.php) instead? [Here is the Quickstart az_card formatter](https://github.com/az-digital/az_quickstart/blob/main/modules/custom/az_card/src/Plugin/Field/FieldFormatter/AZCardDefaultFormatter.php) which builds the link render array.
   - (Update Nov 6, 2023) Adding the Linkit formatter seems unnecessary to me for this PR: [see comment below](#issuecomment-1796469153).

## Related issues

#2102

Linkit issues:

- [File is inserted with absolute URL](https://www.drupal.org/project/linkit/issues/2829173)
- [Linkit Link field: File entities not processed like other entities, resulting in double-encoding](https://www.drupal.org/project/linkit/issues/3348154)
- [Direct URL to media file entity does not work because relative URL does not pass URL path validation](https://www.drupal.org/project/linkit/issues/3354873)

## How to test

Probo site: https://7b5e2985-cedf-4a13-ab91-3a35916c835d--pr-2900.probo.build

1. Upload a PDF from the Content > Media tab.
2. Create a page and add a Cards paragraph.
4. In the Card Link URL field, start typing the name of the PDF. When the file appears in the dropdown below the field, click on it to select it.
5. Save the page and confirm that no errors are displayed.
6. Click the card link and confirm that the PDF is accessed.
7. Edit the page again and test valid and invalid links to other files and Drupal content to confirm that validation still works as expected.

### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
   - [x] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
   - [ ] Adding experimental module
   - [ ] Update experimental module
- **Minor release changes**
   - [ ] New feature
   - [ ] Breaking or visual change to existing behavior
   - [ ] Upgrade experimental module to stable
   - [ ] Enable existing module by default or database update
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- **Other or unknown**
   - [ ] Other or unknown

### Drupal core
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major or minor level update
- **Other or unknown**
   - [ ] Other or unknown

### Drupal contrib projects
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major level update
- **Other or unknown**
   - [ ] Other or unknown

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by tadean** [comment]
The `/sites/default/files/` approach probably will not work for multisite installations. I think we'd need to determine if the actual path is valid for the public file system.

**Comment by bberndt-uaz** [comment]
Next steps:

- Change /sites/default/files to correct path from public://
- Decide if we need to check that the file actually exists

**Comment by bberndt-uaz** [comment]
To test standard Linkit behavior locally, I added a new Link field to AZ Person that used the Linkit formatter for the display and the Linkit widget for the form display (with the "Quickstart LinkIt" profile set for both). In testing links to various nodes and files, I found that [this test in the Linkit formatter was never passed](https://git.drupalcode.org/project/linkit/-/blob/6.1.x/src/Plugin/Field/FieldFormatter/LinkitFormatter.php#L116), so the path validation on the following line was never executed.

With that in mind, I think that using a custom formatter which extends the Linkit formatter would not really provide any benefit. In the future, we could consider having az_card use a Link field and then implement the Linkit formatter at that time. For this PR, I did add a test to the form validation which confirms that any file links on cards do point to files that exist.

File: modules/custom/az_card/src/Plugin/Field/FieldFormatter/AZCardDefaultFormatter.php [hunk 1]
```diff
@@ -9,6 +9,7 @@
 use Drupal\Core\Url;
 use Drupal\paragraphs\ParagraphInterface;
 use Symfony\Component\DependencyInjection\ContainerInterface;
+use Drupal\Core\StreamWrapper\PublicStream;
```
Comment (joeparsons) [inline modules/custom/az_card/src/Plugin/Field/FieldFormatter/AZCardDefaultFormatter.php:12]: We should keep these `use` statements alphabetical (so move this to line 9).

File: modules/custom/az_card/src/Plugin/Field/FieldWidget/AZCardWidget.php [hunk 2]
```diff
@@ -10,6 +10,7 @@
 use Drupal\Core\Url;
 use Symfony\Component\DependencyInjection\ContainerInterface;
 use Symfony\Component\Validator\ConstraintViolationInterface;
+use Drupal\Core\StreamWrapper\PublicStream;
```
Comment (joeparsons) [inline modules/custom/az_card/src/Plugin/Field/FieldWidget/AZCardWidget.php:13]: See comment above about keeping the `use` statements in alphabetical order (should do that here, too).

**Comment by bberndt-uaz** [comment]
@tadean Regarding the use of [str_starts_with()](https://github.com/az-digital/az_quickstart/pull/2900/files#diff-0971e45bcb723eb12cc59af76162069d9bc1fe907c96df692b4dadde7c4a257eR528), I checked locally with a new Link field using Linkit on AZ Person and found that Linkit does not allow links that start with `./`. Linkit displays an error message which says: `Manually entered paths should start with one of the following characters: / ? #`. If you have suggestions for anything else I should test or something I should change the `str_starts_with()` condition to, please let me know!

**Comment by tadean** [comment]
> @tadean Regarding the use of [str_starts_with()](https://github.com/az-digital/az_quickstart/pull/2900/files#diff-0971e45bcb723eb12cc59af76162069d9bc1fe907c96df692b4dadde7c4a257eR528), I checked locally with a new Link field using Linkit on AZ Person and found that Linkit does not allow links that start with `./`. Linkit displays an error message which says: `Manually entered paths should start with one of the following characters: / ? #`. If you have suggestions for anything else I should test or something I should change the `str_starts_with()` condition to, please let me know!

@bberndt-uaz Sorry, the path type I was referring to were ones that contain `..` as a intermediary component of the path, like `/sites/default/files/../some/other/path`

**Comment by bberndt-uaz** [comment]
@tadean I'm sorry, I misunderstood what you were saying during the meeting today. I tried entering the following paths in a "standard" Linkit Link field and in the Card Link URL field in this PR:

Linkit Link field:
- /sites/default/files/../../../favicon-32x32.png: Allowed in the field, linked successfully on the page
- /sites/default/files/../../../themes/contrib/bootstrap_barrio/images/help.png: Allowed in the field, linked successfully on the page
- /sites/default/files/../../../favicon-missing-file.png: Allowed in the field, link leads to "Not Found" error page
- /sites/default/files/../../../pages/text: Allowed in the field, linked successfully on the page

Card Link URL field in this PR:
- /sites/default/files/../../../favicon-32x32.png: Allowed in the field, linked successfully on the page
- /sites/default/files/../../../themes/contrib/bootstrap_barrio/images/help.png: Allowed in the field, linked successfully on the page
- /sites/default/files/../../../favicon-missing-file.png: Not allowed in the field (presumably fails the file_exists() check)
- /sites/default/files/../../../pages/text: Not allowed in the field (presumably fails the file_exists() check)

Do you think any change is needed to account for the potential use of `..` in the path?

**Comment by tadean** [comment]
Hmm...It might be worth testing if a file exists call using the actual stream wrapper might prevent the usage of `..` to "escape" out of the directory...e.g. something like testing `file_exists('public://path/to/my/file.pdf')`

**Comment by bberndt-uaz** [comment]
@tadean Thanks for the suggestion for using the actual stream wrapper! I have updated the `file_exists()` call and I believe that now only links to public files are allowed. Now the four test paths in my previous comment all trigger the "link does not exist" error message. I think this PR is ready for final review!

**Review by trackleft — APPROVED** [review]
Seems to work

**Review by joeparsons — APPROVED** [review]

**Review by tadean — APPROVED** [review]

**Pull request merged**